docs/content/en/highlights/2021-07-21-0-16-upgrade-guide.md

```diff
@@ -44,3 +48,88 @@ inputs = ["datadog_agent"]
 ```
 
 Encoding fields other than `codec` are still valid.
+
+### Vector source/sink version 2 released {#vector-source-sink}
+
+We've released a new major version (`v2`) of our `vector` [source][]/[sink][]
+components. This release resolves a lot of issues we saw with our previous
+(`v1`) TCP-based implementation of these two components.
+
+The new version transitions to using gRPC over HTTP as its communication
+protocol, which resolves those issues.
+
+To allow operators to transition at their leisure, this new release of Vector
+still defaults to `v1`. In the next release (`0.17.0`) we'll require operators
+to explicitly state which version they want to use, but continue to support
+`v1`. The release after that (`0.18.0`) we'll drop `v1` completely, and default
+to `v2`, we also no longer require you to explicitly set the version since there
+will only be one supported going forward.
+
+If you want to opt in to the new (stable!) `v2` version, you can do so as
+follows:
+
+```diff
+[sinks.vector]
+  type = "vector"
++ version = "v2"
+
+[sources.vector]
+  type = "vector"
++ version = "v2"
+```
+
+There are a couple of things to be aware of:
+
+#### Upgrade both the source _and_ sink
+
+You **have** to upgrade **both** the source and sink to `v2`, or none at all,
+you cannot update one without updating the other. Doing so will result in a loss
+of events.
```

Comment:
Will they lose data or just see a bunch of errors?

Reply:
Based on discord reports, silent. There doesn't seem to have been errors reported when there's a version mismatch.

Ex: https://discord.com/channels/742820443487993987/861983462910590986/869662768649867304

Reply:
Hm, @JeanMertz should we try to resolve that as part of releasing this? I would assume we'd get some sort of network error.

Reply:
Yeah, I was surprised by this too. I would think the `vector` source would error. If we could have it do that, I think that would help people in the transition. A couple of people have run into that issue in Discord.

The `vector` sink would really have no way of knowing since it just writes bytes to the TCP socket without any feedback mechanism.

Reply:
So, I looked into this.

If I try to connect a `v1` sink to a `v2` source, the v1 sink will generate "broken pipe" internal events because it can't connect to the source, but it doesn't actually stop Vector, nor does it produce any error messages in the log output.

To the user, it will appear as if the client and server are "hanging", in that the client never sends any events to the server, and if you're not capturing internal events, you'll never get any indication why it's hanging.

We can do three things here:

1. Leave as is
2. Explicitly log warning or error message (note that this would apply to _all_ sinks that rely on our TCP sink implementation, since that's what the old Vector TCP-based sink delegates to)
3. Terminate Vector (I wouldn't advocate for this, but adding it here for completeness)

---

Doing the reverse (i.e. `v2` sink connecting to a `v1` source) _does_ indicate what's happening, both on the sink and source side:


```
Aug 10 12:56:26.364 ERROR vector::topology::builder: msg="Healthcheck: Failed Reason." error=Vector source unhealthy component_kind="sink" component_type="vector" component_name="vector"
Aug 10 12:56:27.377 ERROR sink{component_kind="sink" component_name=vector component_type=vector}:request{request_id=0}: vector::sinks::util::retries: Non-retriable error; dropping the request. error=Request failed: status: Unknown, message: "connection error: broken pipe", details: [], metadata: MetadataMap { headers: {} }
```

```
Aug 10 12:56:35.382  WARN source{component_kind="source" component_name=vector component_type=vector}:connection{peer_addr=127.0.0.1}: vector::sources::util::tcp: Failed to read data from TCP source. error=frame size too big
```

---

Given this, we should focus on the first use-case, and the most straightforward solution would be to log a warning if any TCP sink ever receives an error from the server (or can't connect to the server). _Or_ if we consider having internal events indicate something is going wrong as "good enough", then the first case is also already tackled, as it does trigger an internal event, it's just that you might miss it if you don't log those events.

Reply:
I should clarify, we _do_ emit a log message (`TcpSocketError`), but it's set to `debug`, which means you won't see it unless you explicitly enable debug logging.

Reply:
Aha, I see. I could see bumping that to a warning.

Reply:
Resolved in b3c0ff79cc37a33af5906f471ff63900e45e6140